### DISCUSSION TITLE
Can 2 supported Python Kernel co-exist in one JupyterLite?

### DISCUSSION THREAD

**Discussion by plommon** [jupyterlite/jupyterlite#1379] [Q&A]
Hi guys, I'd like to know if it's possible to have `jupyterlite-pyodide-kernel` and `jupyterlite-xeus` at the same time for one JupyterLite? Thanks in advance for your attention!

**Comment by jtpio** [comment]
@plommon yes absolutely.

Not sure there is an example out there, but you should be able to follow the instructions for installing the two kernels as part of the same JupyterLite deployment.

**Comment by sorphwer** [comment]
@plommon  Theoretically yes but I haven't installed them together after hours of attempting. 

Here is what I got so far:

You must use conda/mamba to install both kernel since `jupyterlite-xeus` require an activated conda environment to install its dependencies in `environment.yml`, therefore you need to have a `build-environment.yml` where you put `jupyterlite-pyodide-kernel` and `jupyterlite-xeus`.

However, I keep getting errors when building jupyterlite under this enviroment, it seems when conducting `jupyter lite build --contents content --output-dir dist --XeusAddon.environment_file=environment.yaml`, some shared library like `python` or `pip` will cause errors but I am still not sure and trying to fix them.

In another words, it looks possible since pyodide itself can be regarded as a big js file and it shouldn't effect xeus, while there are issues during complex dependency linking

**Comment by jtpio** [comment]
Here is test repo that uses both the Pyodide and the Xeus Python kernels: https://github.com/jtpio/lit

The `environment.yml` file is used to specify the dependencies used in the Xeus Python kernel. For Pyodide, it is indeed more of a predefined list of files and packages, which are independent from the packages used by the Pyodide kernel.

![image](https://github.com/jupyterlite/jupyterlite/assets/591645/22902bc7-b0df-47ba-b7e8-032f9b7a35cd)

**Reply by sorphwer** [reply]
Hi @jtpio ,
Thanks for demonstarting this and it looks encouraging! I wonder in what platform setup you just conduct `jupyter lite build` successfully since I still get error using your repo, and I guess there should be something wrong with my system setup, here are more details:

**Platform**: Windows10 64-bit with micromamba and Python 3.9 installed
In my repo I tested:
1. Under venv, use `pip install -r requirements.txt` then  `jupyter lite build`, I got missing mamba/conda.  (That's why I once believed we must use mamba/conda all the time in my early reply)
2. Use `micromamba create -f build-environment.yml`, and `micromamba activate` then `jupyter lite build`, I got error **Cannot find a valid extracted directory cache for 'python-3.11.3-h_hash_24_cpython.tar.bz2'** when micromamba trying to install packages in `environment.yml` 

In [https://github.com/jtpio/lit](https://github.com/jtpio/lit) I tried: 
1. Under venv, use `pip install -r requirements.txt` then  `jupyter lite build`, I got same error **Cannot find a valid extracted directory cache for 'python-3.11.3-h_hash_24_cpython.tar.bz2'**

I believe it will be very helpful if you could share your platform setup and commands to build juputerlite, or I wonder if you just run it in Github Actions
